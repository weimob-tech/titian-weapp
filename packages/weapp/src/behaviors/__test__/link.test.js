const path = require('path');
const { getBehavior } = require('../../common/test/getBehavior');
const { getComponent } = require('../../common/test/getComponent');

describe('test behavior link', () => {
  const stack = [];
  const { getCurrentPages } = global;

  beforeAll(() => {
    global.getCurrentPages = jest.fn(() => stack);
  });

  afterAll(() => {
    global.getCurrentPages = getCurrentPages;
  });

  it('test link navigateTo, props success', (done) => {
    const behavior = getBehavior(path.join(__dirname, '../link.ts'));
    const mockNavigateToFn = jest.fn();
    const mockRedirectToFn = jest.fn();
    const { navigateTo, redirectTo } = global.wx;

    global.wx.navigateTo = () => {
      stack.push('navigateTo');
      mockNavigateToFn();
    };

    global.wx.redirectTo = async () => {
      stack.push('redirectTo');
      return mockRedirectToFn();
    };

    const comp = getComponent(
      {
        template: '<view class="test" bind:tap="jumpToLink">{{link}}</view>',
        behaviors: [behavior]
      },
      {
        link: 'pages/root/index'
      }
    );

    for (let i = 0; i <= 10; i += 1) {
      comp.querySelector('.test').dispatchEvent('tap');
    }

    Promise.resolve().then(() => {
      expect(comp).toMatchSnapshot();
      expect(mockNavigateToFn).toBeCalledTimes(10);
      expect(mockRedirectToFn).toBeCalledTimes(1);
      expect(stack).toEqual([
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'navigateTo',
        'redirectTo'
      ]);
      global.wx.navigateTo = navigateTo;
      global.wx.redirectTo = redirectTo;
      done();
    });
  });
});
