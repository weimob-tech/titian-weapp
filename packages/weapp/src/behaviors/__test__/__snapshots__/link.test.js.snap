// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test behavior link test link navigateTo, props success 1`] = `
<main>
  <wx-view
    class="test"
    bind:tap="jumpToLink"
  >
    pages/root/index
  </wx-view>
</main>
`;
